src/execution_plan/optimizations/utilize_indices.c

```diff
@@ -19,136 +23,411 @@ AST_Operator _reverseOp(AST_Operator op) {
+				break;
+			case OP_LE:    // <=
+				node = RediSearch_CreateLexRangeNode(sp, field, RSLEXRANGE_NEG_INF, v.stringval, 0, 1);
+				break;
+			case OP_GT:    // >
+				node = RediSearch_CreateLexRangeNode(sp, field, v.stringval, RSLECRANGE_INF, 0, 0);
+				break;
+			case OP_GE:    // >=
+				node = RediSearch_CreateLexRangeNode(sp, field, v.stringval, RSLECRANGE_INF, 1, 0);
+				break;
+			case OP_EQUAL:  // ==
+				node = RediSearch_CreateTokenNode(sp, field, v.stringval);
+				break;
+			case OP_NEQUAL: // !=
+				node = RediSearch_CreateNotNode(sp);
+				RediSearch_QueryNodeAddChild(node, RediSearch_CreateTokenNode(sp, field, v.stringval));
+				break;
+			default:
+				assert("unexpected operation");
+			}
+			break;
+
+		case T_DOUBLE:
+		case T_INT64:
+		case T_BOOL:
+			d = SI_GET_NUMERIC(v);
+			switch(filter->pred.op) {
+			case OP_LT:    // <
+				node = RediSearch_CreateNumericNode(sp, field, d, RSRANGE_NEG_INF, false, false);
+				break;
+			case OP_LE:    // <=
+				node = RediSearch_CreateNumericNode(sp, field, d, RSRANGE_NEG_INF, true, false);
+				break;
+			case OP_GT:    // >
+				node = RediSearch_CreateNumericNode(sp, field, RSRANGE_INF, d, false, false);
+				break;
+			case OP_GE:    // >=
+				node = RediSearch_CreateNumericNode(sp, field, RSRANGE_INF, d, false, true);
+				break;
+			case OP_EQUAL:  // ==
+				node = RediSearch_CreateNumericNode(sp, field, d, d, true, true);
+				break;
+			case OP_NEQUAL: // !=
+				node = RediSearch_CreateNotNode(sp);
+				RediSearch_QueryNodeAddChild(node, RediSearch_CreateNumericNode(sp, field, d, d, true, true));
+				break;
+			default:
+				assert("unexpected operation");
+			}
+			break;
+		default:
+			assert("unexpected value type");
+		}
+	} else {
+		assert("unknow filter tree node type");
+	}
+	return node;
+}
+
+//------------------------------------------------------------------------------
+
+/* Tests to see if given filter tree is a simple predicate
+ * e.g. n.v = 2
+ * one side is variadic while the other side is constant. */
+bool _simple_predicates(const FT_FilterNode *filter) {
+	if(filter->t == FT_N_PRED) {
+		if(filter->pred.lhs->type == AR_EXP_OP || filter->pred.rhs->type == AR_EXP_OP) {
+			return false;
+		}
+
+		// Both left and right side are variadic.
+		if(filter->pred.lhs->operand.type == AR_EXP_VARIADIC &&
+		   filter->pred.rhs->operand.type == AR_EXP_VARIADIC) {
+			return false;
+		}
+
+		// Both left and right are constants.
+		if(filter->pred.lhs->operand.type == AR_EXP_CONSTANT &&
+		   filter->pred.rhs->operand.type == AR_EXP_CONSTANT) {
+			return false;
+		}
+
+		// Validate constant type.
+		SIValue c = SI_NullVal();
+		if(filter->pred.lhs->operand.type == AR_EXP_CONSTANT) c = filter->pred.lhs->operand.constant;
+		if(filter->pred.rhs->operand.type == AR_EXP_CONSTANT) c = filter->pred.rhs->operand.constant;
+		SIType t = SI_TYPE(c);
+		if(!((t & SI_NUMERIC) | (t & SI_STRING) | (t & T_BOOL))) {
```

Comment:
I feel like this is a more intuitive way to build and test against the bitmask:
`if(!(t & (SI_NUMERIC | SI_STRING | T_BOOL)))`
But don't have a strong preference.

Reply:
Agree, much cleaner